client_it_test.go

```diff
@@ -92,6 +95,163 @@ func TestClientLifecycleEvents(t *testing.T) {
 	})
 }
 
+func TestClient_AddLifecycleListener(t *testing.T) {
+	receivedStates := []hz.LifecycleState{}
+	receivedStatesMu := &sync.RWMutex{}
+	it.Tester(t, func(t *testing.T, client *hz.Client) {
+		subscriptionID, err := client.AddLifecycleListener(func(event hz.LifecycleStateChanged) {
+			receivedStatesMu.Lock()
+			defer receivedStatesMu.Unlock()
+			// client is already connected
+			switch event.State {
+			case hz.LifecycleStateShuttingDown:
+				t.Logf("Received shutting down state")
+			case hz.LifecycleStateShutDown:
+				t.Logf("Received shut down state")
+			default:
+				t.Log("Received unknown state:", event.State)
+			}
+			receivedStates = append(receivedStates, event.State)
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		require.NotEqual(t, types.UUID{}, subscriptionID, "subscription UUID should not be empty")
+		if err = client.Shutdown(context.Background()); err != nil {
+			t.Fatal(err)
+		}
+		defer func() {
+			receivedStatesMu.Lock()
+			receivedStates = []hz.LifecycleState{}
+			receivedStatesMu.Unlock()
+		}()
+		targetStates := []hz.LifecycleState{
+			hz.LifecycleStateShuttingDown,
+			hz.LifecycleStateShutDown,
+		}
+		it.Eventually(t, func() bool {
+			receivedStatesMu.RLock()
+			defer receivedStatesMu.RUnlock()
+			return reflect.DeepEqual(targetStates, receivedStates)
+		})
+	})
+}
+
+func TestClient_RemoveLifecycleListener(t *testing.T) {
+	var lifecycleEventReceived int32 = 1
```

Comment:
Why do you set `lifecycleEventReceived = 1` and use `atomic.SwapInt32` to modify it? If there's not an important reason, could you use `atomic.AddInt32` like the rest of the codebase?
```
var lifecycleEventReceived int32
// ...
atomic.AddInt32(&lifecycleEventReceived, 1)
```

Reply:
There isn't any important reason, fixed: [88b6f60](https://github.com/hazelcast/hazelcast-go-client/pull/809/commits/88b6f60a915c3d0237ae5f1fb97bdbda52cedfbd)